### PR TITLE
Use group_interchangeable_qubits for qubit comparison in GateOperation

### PR DESCRIPTION
**Pull request by Ashalynd**
Fix for issue #4766 
Also, Introduced more idiomatic code for group_interchangeable_qubits implementation.

### CODE REVIEW

File: cirq-core/cirq/ops/gate_operation.py [hunk 1]
```diff
@@ -146,17 +146,18 @@ def _json_dict_(self) -> Dict[str, Any]:
...
-            groups[k].append(q)
-        return tuple(sorted((k, frozenset(v)) for k, v in groups.items()))
+        else:
+            return tuple(
+                (k, frozenset(g for _, g in kg))
+                for k, kg in itertools.groupby(
+                    enumerate(self.qubits),
+                    lambda i_q: cast(
+                        gate_features.InterchangeableQubitsGate, self.gate
+                    ).qubit_index_to_equivalence_group_key(i_q[0]),
```
Comment (MichaelBroughton) [inline cirq-core/cirq/ops/gate_operation.py:158]: Could we break this up into a few simpler lines ? This is a little hard to read.

**Review by MichaelBroughton — COMMENTED** [review]
Looks, good just a little readability cleanup then we should be good to merge.

File: cirq-core/cirq/ops/gate_operation.py (hunk 1 above)
Comment (Ashalynd) [reply cirq-core/cirq/ops/gate_operation.py]: Done.

**Review by MichaelBroughton — APPROVED** [review]
LGTM

**Pull request merged**